src/logics/ArithLogic.cc

```diff
@@ -669,16 +676,32 @@ PTRef ArithLogic::mkTimes(vec<PTRef> && args) {
     SimplifyConstTimes simp(*this);
     args.clear();
     SymRef s_new;
-    simp.simplify(getTimesForSort(returnSort), flatten_args, s_new, args);
-    PTRef tr = mkFun(s_new, std::move(args));
-    // Either a real term or, if we constructed a multiplication of a
-    // constant and a sum, a real sum.
-    if (isNumTerm(tr) || isPlus(tr) || isUF(tr) || isIte(tr))
-        return tr;
-    else {
-        auto termStr = pp(tr);
-        throw LANonLinearException(termStr.c_str());
+    simp.simplify(getTimesLinForSort(returnSort), flatten_args, s_new, args);
+    if (!isTimes(s_new)) return mkFun(s_new, std::move(args));
+    PTRef coef = PTRef_Undef;
+    std::vector<PTRef> vars;
+    //    return mkFun(s_new, std::move(args));
+    // Splitting Multiplication into constant and variable subterms
+    for (int i = 0; i < args.size(); i++) {
+        if (isConstant(args[i])) {
+            assert(coef == PTRef_Undef);
+            coef = args[i];
+            continue;
+        }
+        vars.push_back(args[i]);
     }
+    assert(!vars.empty());
+    PTRef tr;
+    if (vars.size() > 1) {
+        if (coef == PTRef_Undef) {
+            tr = mkFun(getTimesNonlinForSort(returnSort), vars);
+        } else {
+            tr = mkFun(s_new, {coef, mkFun(getTimesNonlinForSort(returnSort), vars)});
+        }
+    } else {
+        tr = mkFun(s_new, {coef, vars[0]});
+    }
+    return tr;
```

Comment:
Note to myself. It would be good if we can avoid this complicated piece of code. It feels like the code in `simplify` should take care of most of this.

Reply:
Ideally `simplify` would order args in some way
(For example, if there is a constant, it would be always in `new_args[0]`)

Then, indeed, pretty much everything can be handled with `simplify`...